test/embedding/embeddingdl.c

```diff
@@ -0,0 +1,198 @@
+
+#ifdef _OS_WINDOWS_
+#include <windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+// Load a minimal subset of the functions from julia.h dynamically and
+// masquerade the function pointers to look like the original
+// functions.
+
+// We could get the definitions of these structs from
+// `#include "julia.h"`, but we only need to declare them and
+// including julia.h would interfere with the masquerading trick.
+typedef struct _jl_value_t jl_value_t;
+typedef struct _jl_sym_t jl_sym_t;
+typedef struct _jl_module_t jl_module_t;
+
+// Declare pointers to the functions we need to load from libjulia.
+// Obviously these signatures must match the corresponding functions
+// in the julia.h that libjulia was built from.
+static void (*p_jl_init)(void);
+static int (*p_jl_is_initialized)(void);
+static void (*p_jl_atexit_hook)(int);
+static jl_value_t *(*p_jl_eval_string)(const char *);
+static jl_sym_t *(*p_jl_symbol)(const char *);
+static jl_module_t **p_jl_main_module;
+static jl_value_t *(*p_jl_get_global)(jl_module_t *, jl_sym_t *);
+static void *(*p_jl_unbox_voidpointer)(jl_value_t *);
+static const char *(*p_jl_string_ptr)(jl_value_t *);
+static jl_value_t *(*p_jl_exception_occurred)(void);
+
+// Helper function to extract function pointers from the dynamically
+// loaded libjulia.
+#ifdef _OS_WINDOWS_
+static void *load_function(HMODULE libjulia, const char *name, int *success)
+#else
+static void *load_function(void *libjulia, const char *name, int *success)
+#endif
+{
+#ifdef _OS_WINDOWS_
+    void *p = GetProcAddress(libjulia, name);
+#else
+    void *p = dlsym(libjulia, name);
+#endif
+
+    // Unfortunately Julia renames jl_init to jl_init__threading if
+    // Julia is compiled with threading support, so we have to check
+    // which of these is available, or otherwise query Julia in some
+    // other way (issue #28824).
+    if (!p && strcmp(name, "jl_init") == 0) {
+#ifdef _OS_WINDOWS_
+        p = GetProcAddress(libjulia, "jl_init__threading");
+#else
+        p = dlsym(libjulia, "jl_init__threading");
+#endif
+    }
+
+    if (!p) {
+        fprintf(stderr, "%s not found in libjulia.\n", name);
+        *success = 0;
+    }
+
+    return p;
+}
+
+// Open libjulia and extract pointers to the needed functions.
+static int load_julia_functions()
+{
+#ifdef _OS_WINDOWS_
+    // libjulia.dll needs to be in the same directory as the
+    // executable or in PATH.
+    const char *library_name = "libjulia.dll";
+    HMODULE libjulia = LoadLibrary(library_name);
+#else
+    // libjulia.so needs to be in LD_LIBRARY_PATH. It could also be in
+    // rpath (but that kind of defeats the purpose of dynamically
+    // loading libjulia) or an absolute path could be given, computed
+    // from other information.
+#ifdef __APPLE__
+    const char *library_name = "libjulia.dylib";
+#else
+    const char *library_name = "libjulia.so";
+#endif
+    void *libjulia = dlopen(library_name, RTLD_LAZY | RTLD_GLOBAL);
+#endif
+
+    if (!libjulia) {
```

Comment:
Might be nice to also do some version checking here (`jl_ver_major`, at least?).

Though I don't know whether the embedding API has any ABI guarentees yet?

Reply:
To clarify what I mean about ABI guarantees — those would define how closely we must match `jl_ver_major` / `jl_ver_minor` / `jl_ver_patch` in order to further use the `libjuila` that we are `dlopen`ing.

Reply:
Keno comments on slack that "there are no stability guarantees on libjulia".

So naively we should check for a match of `JULIA_VERSION_MAJOR == jl_ver_major()` and the same for `jl_ver_minor` at least. Maybe also `jl_ver_patch` though I hazard a guess that a plugin based on the functions in jlapi.c is likely to keep working between patch versions.

All of this can be put into an inline `jl_check_version()` function.

Reply:
Even more useful would be to have version number information available in global variables, which could be checked before calling `jl_init`/`jl_init_with_image` in case those interfaces should change or if you use custom system images, which are inherently version dependent.
Cf #32614.